### DISCUSSION TITLE
Custom AXEvent

### DISCUSSION THREAD

**Discussion by rywager** [AxisCommunications/acap-native-sdk-examples#226] [Q&A] [answered]
Hello,

For my app i would like to be able to create a custom event and have it declared and available for rules and the MQTT publications.

In the native example, the only reference to this is that axis uses "tnsaxis" topic for theirs.

Few questions

1. I was wondering if anyone had any examples where a completely custom event could be created and have it permiate through the axis event system?
2. For the token, do i just pick any 4 digit number?
3. My events would want to involve 4 different pieces of data, is there a way to send a struct/json as part of the event or do i break this up across multiple?
4. Once this is done, will the demo get_eventlist.py setup see my event or will i have to customize it in someway?

Thanks!

**Comment by Cacsjep** [comment, marked as answer]
> 1. I was wondering if anyone had any examples of creating a completely custom event that could permeate through the AXIS event system?

The [Send Event example](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/axevent/send_event) provides all the information you need.

You can also check out these [SDK page examples](https://axiscommunications.github.io/acap-documentation/3.5/api/axevent/html/examples.html).

> 2. For the token, do I just pick any 4-digit number?

You can use whatever you want; this is just an example value, e.g., a reference for the event.

> 3. My events would involve 4 different pieces of data. Is there a way to send a struct/JSON as part of the event, or do I break this up across multiple entries?

I found it easier to use multiple entries, but you could check out [AXEventElementItem](https://axiscommunications.github.io/acap-documentation/3.5/api/axevent/html/ax__event__element__item_8h.html). You could also store a JSON in a string entry and parse it afterward. AXIS did this as well with SVG data in the ObjectAnalytics app like `<tt:SimpleItem Name="svgframe" Value="<svg><desc></desc></svg>"/>`.

> 4. Once this is done, will the demo `get_eventlist.py` setup see my event, or will I have to customize it in some way?

Yes, it will.

I suppose you would also select your event in the ui therefore you should create the event with tnsaxis and CameraApplicationPlatform as here in this [example](https://axiscommunications.github.io/acap-documentation/3.5/api/axevent/html/ax_event_stateless_declaration_example_8c-example.html)

**Reply by rywager** [reply]
1.  Yes it's a good example but from what i see there is tns1 for ONVIF, tnsaxis for axis defined.  I wasn't sure if i could just use tnsXXX or if i needed to use the tnsaxis for example. 
2. Perfect thanks.
3. i will check that out thanks!
4. Awesome, i'll be working on this today and report back with findings.  Much appreciated.

**Reply by Cacsjep** [reply]
if u are familiar with golang check the code below. use tnsaxis because CameraApplicationPlatform events are tnsaxis namespaced

**Reply by rywager** [reply]
Just deleted my follow up comment because i found what i had been ignoring.  Once you use the topic0 tnsaxis of literally CameraApplicationPlatform everything else fell into place.  I don't know why up until now i thought that was just a placeholder in the examples :) Thanks so much!!!

   ax_event_key_value_set_add_key_values(key_value_set,
        NULL,
        "topic0", "tnsaxis", "CameraApplicationPlatform", AX_VALUE_TYPE_STRING,
        "topic1", "tnsaxis", "Baton", AX_VALUE_TYPE_STRING,
        "event", NULL, NULL, AX_VALUE_TYPE_BOOL,
        NULL);

**Reply by Cacsjep** [reply]
Ok, great, I fell for it the first time too =) 
If you have more questions, you can ask them on Discrod, that would help expand the Discord [channel ](https://discord.gg/UtdzzrGf) 😀

**Comment by Cacsjep** [comment]
I dont know if you are familiar with golang but i create a [example ](https://github.com/Cacsjep/goxis/tree/master/examples/axevent/sendMqtt) for you that shows how you could do this.

```go
package main

import (
	"math/rand"
	"time"

	"github.com/Cacsjep/goxis/pkg/acapapp"
	"github.com/Cacsjep/goxis/pkg/axevent"
	"github.com/Cacsjep/goxis/pkg/utils"
)

func main() {
	// Initialize a new ACAP application instance.
	app := acapapp.NewAcapApplication()

	// Define the structure and metadata for an MQTT event. This includes specifying the event name,
	// a human-readable nice name, and the key-value structure that the event will contain.
	mqttEvent := &acapapp.CameraPlatformEvent{
		Name:     "MqttEvent",                               // Unique identifier for the event.
		NiceName: utils.NewStringPointer("Mqtt Event Send"), // Human-readable name for the event.
		KvsEntries: []*axevent.KeyValueEntrie{ // Define the keys and their value types that this event will include.
			{Key: "foo", ValueType: axevent.AXValueTypeInt},
			{Key: "bar", ValueType: axevent.AXValueTypeDouble},
			{Key: "baz", ValueType: axevent.AXValueTypeString},
			{Key: "qux", ValueType: axevent.AXValueTypeBool},
		},
		DataMarkers: []*axevent.AxEventKeyValueSetDataMark{ // Mark keys that represent event data.
			{Key: "foo"},
			{Key: "bar"},
			{Key: "baz"},
			{Key: "qux"},
		},
		NiceNames: []*axevent.AxEventKeyValueSetNiceNames{ // Assign 'nice names' for keys, improving readability in UI.
			{Key: "foo", KeyNiceName: utils.NewStringPointer("Foo Value")},
			{Key: "bar", KeyNiceName: utils.NewStringPointer("Bar Value")},
			{Key: "baz", KeyNiceName: utils.NewStringPointer("Baz Value")},
			{Key: "qux", KeyNiceName: utils.NewStringPointer("Qux Value")},
		},
		Stateless: true, // Indicates that the event does not maintain any state between emissions.
	}

	// Register the MQTT event with the application. This step declares the event structure to the system,
	// allowing it to recognize and properly route the event when sent.
	mqtt_event_id, err := app.AddCameraPlatformEvent(mqttEvent)
	if err != nil {
		app.Syslog.Critf("Error adding event declaration: %s", err.Error()) // Log critical error if the event declaration fails.
	}

	// Periodically send the MQTT event every second. This goroutine simulates an ongoing event emission,
	// like sensor readings or status updates.
	go func() {
		for true {
			time.Sleep(1 * time.Second) // Wait for 1 second before sending the next event.

			// Attempt to send a newly created MQTT event with dynamic values.
			if err := app.SendPlatformEvent(mqtt_event_id, func() (*axevent.AXEvent, error) {
				// Dynamically generate event data with random values.
				return mqttEvent.NewEvent(acapapp.KeyValueMap{
					"foo": rand.Int(),          // Random integer value.
					"bar": rand.Float64(),      // Random floating-point value.
					"baz": "i want to be json", // Static string value.
					"qux": rand.Intn(2) == 1,   // Random boolean value (true or false).
				})
			}); err != nil {
				app.Syslog.Errorf("Error sending event: %s", err.Error()) // Log error if event sending fails.
			} else {
				app.Syslog.Infof("Event send") // Log information on successful event send.
			}
		}
	}()

	// Execute the main loop of the application. This call blocks and allows the application to continue running,
	// responding to system events, and handling the periodic event emissions initiated above.
	app.Run()
}
```  
This is how it looks in WEBUI
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/98898150/77e9afe0-8e25-4d65-8f2c-9a5bb2816baa)

In Metadata monitor looks like this
![image](https://github.com/AxisCommunications/acap-native-sdk-examples/assets/98898150/ed9f452d-16b0-4541-baf5-8386d68a5fc5)
